src/webgpu/api/validation/copyTextureToTexture.spec.ts

```diff
@@ -0,0 +1,277 @@
+    {
+      srcMipLevelCount: 3,
+      dstMipLevelCount: 3,
+      srcCopyMipLevel: 0,
+      dstCopyMipLevel: 4,
+      _isSuccess: false
+    },
+  ] as const)
+  .fn(async t => {
+    const {
+      srcMipLevelCount,
+      dstMipLevelCount,
+      srcCopyMipLevel,
+      dstCopyMipLevel,
+      _isSuccess
+    } = t.params;
+
+    const srcTexture = t.device.createTexture({
+      size: { width: 32, height: 32, depth: 1 },
+      format: 'rgba8unorm',
+      usage: GPUTextureUsage.COPY_SRC,
+      mipLevelCount: srcMipLevelCount,
+    });
+    const dstTexture = t.device.createTexture({
+      size: { width: 32, height: 32, depth: 1 },
+      format: 'rgba8unorm',
+      usage: GPUTextureUsage.COPY_DST,
+      mipLevelCount: dstMipLevelCount,
+    });
+
+    t.TestCopyTextureToTexture(
+      { texture: srcTexture, mipLevel: srcCopyMipLevel },
+      { texture: dstTexture, mipLevel: dstCopyMipLevel },
+      { width: 1, height: 1, depth: 1 },
+      _isSuccess
+    );
+  });
+
+
+g.test('texture_usage')
+  .params(
+    params()
+      .combine(poptions('srcUsage', kTextureUsages))
+      .combine(poptions('dstUsage', kTextureUsages))
+  )
+  .fn(async t => {
+    const { srcUsage, dstUsage } = t.params;
+
+    const srcTexture = t.device.createTexture({
+      size: { width: 4, height: 4, depth: 1 },
+      format: 'rgba8unorm',
+      usage: srcUsage,
+    });
+    const dstTexture = t.device.createTexture({
+      size: { width: 4, height: 4, depth: 1 },
+      format: 'rgba8unorm',
+      usage: dstUsage,
+    });
+
+    const isSuccess =
+      srcUsage === GPUTextureUsage.COPY_SRC && dstUsage === GPUTextureUsage.COPY_DST;
+
+    t.TestCopyTextureToTexture(
+      { texture: srcTexture },
+      { texture: dstTexture },
+      { width: 1, height: 1, depth: 1 },
+      isSuccess
+    );
+  });
+
+g.test('sample_count')
+  .params([
+    { srcSamplecount: 1, dstSampleCount: 1, _isSuccess: true },
+    { srcSamplecount: 1, dstSampleCount: 4, _isSuccess: false },
+    { srcSamplecount: 4, dstSampleCount: 1, _isSuccess: false },
+    { srcSamplecount: 4, dstSampleCount: 4, _isSuccess: true },
+  ] as const)
+  .fn(async t => {
+    const { srcSamplecount, dstSampleCount, _isSuccess } = t.params;
+
+    const srcTexture = t.device.createTexture({
+      size: { width: 4, height: 4, depth: 1 },
+      format: 'rgba8unorm',
+      usage: GPUTextureUsage.COPY_SRC,
+      sampleCount: srcSamplecount,
+    });
+    const dstTexture = t.device.createTexture({
+      size: { width: 4, height: 4, depth: 1 },
+      format: 'rgba8unorm',
+      usage: GPUTextureUsage.COPY_DST,
+      sampleCount: dstSampleCount,
+    });
+
+    t.TestCopyTextureToTexture(
+      { texture: srcTexture },
+      { texture: dstTexture },
+      { width: 4, height: 4, depth: 1 },
+      _isSuccess
+    );
+  });
+
+g.test('multisampled_copy_restrictions')
+  .params(
+    params()
+      .combine(
+        poptions('srcCopyOrigin', [
+          { x: 0, y: 0, z: 0 },
```

Comment:
nit: also have a test for origin = 0, 0, 0 but the width / height / depth not being the full subresource size (and also at a miplevel to check the miplevel size is correctly used)

Reply:
The mipmap level count of multisampled textures must be 1 so we shouldn't consider mipLevel > 0 here.

Reply:
Done